capi/geos_c.h.in

```diff
@@ -1244,11 +1244,49 @@ extern char GEOS_DLL GEOSisValidDetail_r(
     char** reason,
     GEOSGeometry** location);
 
+/* ========== Make Valid ========== */
+
+/**
+* Algorithm to use when repairing invalid geometries.
+*
+* \see GEOSMakeValidWithOptions
+*/
+enum GEOSMakeValidMethods {
+    /** Original method, combines all rings into
+        a set of noded lines and then extracts valid
+        polygons from that linework. */
+    GEOS_MAKE_VALID_ORIGINAL,
+    /** Buffered method, first makes all rings valid
+        then merges shells and subtracts holes from
+        shells to generate valid result. Similar to
+        using buffer(0) but with common failure
+        cases handled correctly. */
+    GEOS_MAKE_VALID_BUFFERED
+};
+
+/**
+* Options structure to fill out when passing options
+* to GEOSMakeValidWithOptions.
+*/
+typedef struct {
```

Comment:
Does this need to be defined in the implementation file only so that the layout is defined by the library we link against rather than the header we built against?

Reply:
I think it needs to be in the header, or the user can't form the right struct to pass in. I think I may have over-thought this though, copying the GDAL api for grid interpolation, which has 5 different algorithms each of which is several distinct options. But here I have two algorithms and only one currently has an option. A single struct with all the possible options makes more sense, and can avoid the void* in the call.

Reply:
Is `GEOSBufferParams` a pattern that can be copied?  Pretty verbose tho...

Reply:
There's a level of conditionality in this (if algorithm A then options A, if B then options for B) that doesn't exist in the buffer case (which might have been better handled in named parameters, frankly, precisely because the options are fixed, but named parameters didn't exist at the time). I'm probably going to write up something a little more general for this form of options string like the GDAL CSList so at least they can share some utility code.

Reply:
That's a point.  Although all the optionality exists on the `GeoemetryFixer` side, and at the moment `MakeValid` gets the no-option signature.  So the options can just apply to `GeometryFixer`.  If and when the default is changed, then we can add an option that indicates "use the MakeValid algorithm".

Reply:
I tried to keep structs out of that header as much as possible. Opaque void pointers are what made it possible so far to avoid breaking the ABI

Reply:
Ug. Makes me want to just have two function end points, less cluttered than having a whole new options object with creators/destructors/setters/getters for what is, currently, two options (algorithm and keepCollapsed).

Reply:
It's the price of a C API, I think.